### ISSUE TITLE
Compression

### ISSUE THREAD

**Issue by Mitmischer** [xournalpp/xournalpp#1869]
**Is your feature request related to a problem? Please describe.**
Saving a reasonably sized notebook (20 pages) takes several seconds and results in files that seem quite large. There was no discussion of the compression algorithm in use until now.

**Describe the solution you'd like**
I took the largest Xournal file I produced and ran the squash benchmark (https://github.com/quixdb/squash-benchmark) on it. The original xopp file was 38918125 bytes uncompressed. Those are my results:
[xopp.csv.txt](https://github.com/xournalpp/xournalpp/files/4474647/xopp.csv.txt) (it's a csv; file extension to make Github happy) 

For a visualization, you can use the squash web benchmark. As a quick hack, you can copy xopp.csv (from above) to data/hoplite.csv ( or any other machine - just remember the name) and add

```
 { id: 'xopp',
      source: '',
      sourceUrl: '',
      description: 'Xournal file',
      size: 38918125 },
```

to `datasets` (line 71) in `squash-benchmark.js`. Fire up a web server in the squash-benchmark-web directory, a simple python script will do (https://medium.com/@elijahlopezz/python-3-x-simplehttpserver-7cf0bbe3a798). The result looks like this: https://quixdb.github.io/squash-benchmark/. Of course, their data is different. Select xopp and hoplite on your local version to see my data. Here's the most important graph:
![image](https://user-images.githubusercontent.com/774380/79220258-fb328980-7e53-11ea-9a7b-47fd1ac2a358.png)

It looks like zstd uses gzip with level 6 by default (?). That gives a compression ratio of 3.02 with a speed of 16.57 MiB/s (which equates to the notebook taking about 2.5 seconds to save, feels right) I was surprised that we can't get significantly better than this without getting very slow. brotli on level 4 achieves 3.56 with 16.45 MiB/s and zstd on level 7 gets 3.33 with 18.93 MiB/s, getting better compression with comparable speed. The room for improvements is obvious but small. zstd decompresses faster than zlib(gzip) among all configurations, brotli is comparable but has more extreme extreme values. Optimizing for speed (while keeping the ratio!), one can just use brotli or zstd on level 1. They give 3.28, 76.57 MiB/s and 3.13, 78.79 MiB/s, respectively. Note that squash has not been updated in a long time (last commit on Dec 2018), so results are inherently outdated. I ran the test on current zstd (using its own benchmark mode) and got comparable results, however.

My conclusion from that data is: We can't really compress much better than we currently do - realistically, we can shave off maybe 15% from file size. But we can compress ~5 times as fast as currently, decreasing saving times significantly, using either brotli or zstd. As honestly the current state is good enough, this might not be an important optimization. But the data could be useful in discussion of the new file format.

I am open to your questions and willing to run the benchmark again on different notebooks (if that helps)

**Describe alternatives you've considered**
#1277

**Comment by Technius** [comment]
Thanks for the great analysis! The primary problem, as I see it, is that the annotations are stored in XML format rather than as binary data. There are a lot of tags and each tag consists of multiple characters when we could just replace all of the tag characters with a 1 or 2 byte id. We could introduce a binary encoding for the new file format, which will lead to significant space savings but simultaneously make it more error-prone/harder to debug.

**Comment by Mitmischer** [comment]
You're welcome!
I also thought that Xournal would profit from a binary format.. but with all things considered, that's not a straightforward decision to make! Looking at the savings in size:
The big contributors to file size are the strokes - all other tags weigh significantly less.
Every stroke needs an identifier and it has pen, ts, fn and color. That are 1+1+4+1(?)+3 bytes. I assumed 4 bytes for ts and one byte for fn, totalling 10 bytes per stroke. The width attribute and the text of the stroke will contribute 8 bytes per item, saving the double in full precision. I'd argue to use floats here (4 bytes per item), as that would give huge savings with only a small loss in precision..? Leaving that open to discussion ;)
With those assumptions, a hypothetical binary format would have the following footprint ( https://gist.github.com/Mitmischer/2d1ce097a5d8d68cbe25f79ffac71536 ):

```
30190 strokes, totalling 294.8 KiB
1004785 elements in width attributes, totalling 7.7 MiB
2011044 coordinates, totalling 15.3 MiB
total size: 23.3 MiB (24428532 bytes)
```
That would be 60% of the XML file size. The compressed XML, however, is still a lot smaller (~13MiB)! It would be interesting to compress a binary version of the Xournal notebook and compare it against the compressed XML. I learned that binary formats don't usually respond well to compression, so that's quite a bummer. Maybe we have enough repetition/similarity to make compression worthwhile, but I'm not sure. Saving the coordinates as floats will squeeze the file to ~11.7MiB, which is (without compression!) less than the compressed XML.

So, what do we get from this? :) The XML tags don't add a lot of weight. It's all about storing the coordinates themselves efficiently, or reducing their count or precision!

**Comment by Technius** [comment]
> The big contributors to file size are the strokes - all other tags weigh significantly less.

You're right--I had forgotten what the actual encoding of the strokes were.

> I also thought that Xournal would profit from a binary format.. but with all things considered, that's not a straightforward decision to make!

I've been thinking about this lately and I don't think it would be too much different. For the new file format, we could serialize annotation data using [MessagePack](https://msgpack.org/index.html), which is basically like binary JSON. The serialization/deserialization code would basically be the same thing that we have now, except safer because we aren't manually constructing XML strings.

**Comment by LittleHuba** [comment]
The new format will actually use [Cap'nProto](https://capnproto.org) which is extremely fast and does not introduce a lot of overhead. 
In essence there is no searialization step since the data is already stored in memory in a way where you can write it out and save it to a file directly. Check out their page for details.

**Comment by Mitmischer** [comment]
Sounds good, looking forward to it?
Do you need help with the implementation?

**Comment by LittleHuba** [comment]
I'll upload the current code tomorrow to a repo of our team. Most classes for management are implemented but we will need to move quite some code from X++ to the library. You are more then welcome to help with that. 

But first we need a code review and testing of what I cooked up so far. You can of course give your two cents there as well.

**Comment by Mitmischer** [comment]
I have some time at hand at the moment, so I'm happy to help! I guess I better start taking a look at the documentation ;)

**Comment by LittleHuba** [comment]
You could check out how Gtest works. Well need tons of testing for the code as the next step before we can really implement anything.

**Comment by Mitmischer** [comment]
:+1: I've been using GTest in a project before! Easy to use and it just works. I also set up CI for the other project, I could do that here, too (I don't have a reliable server at hand, though)

**Comment by LittleHuba** [comment]
Can you PM me on gitter? I'll give you some work then 😉

**Comment by Technius** [comment]
For now, I think we can switch to [`zstd`](https://github.com/facebook/zstd) as mentioned. It's BSD / GPLv2 and widely available, so it should basically be a drop-in replacement for zlib.
